### PR TITLE
Dev-Container support

### PR DESCRIPTION
**Pull request by jonah-iden**
#### What it does

This implements a first version of the [Dev-Container Support](https://code.visualstudio.com/docs/devcontainers/containers) vscode offers for Theia. This can later also be used to forward arbritrary ports at container runtime

The [devcontainer.json format](https://containers.dev/implementors/json_reference/) offers alot of configuration option of which currently only a certain few are implemented.

Implemented properties are:
- `image`
- `dockerfile`/`build.dockerfile` (dockerfile support still needs some more testing and work regarding its options)
  - including `build.args` which are just passed to docker as is
- `forwardPorts`
- `mounts`

The communication with the theia backend inside the container is done through a docker exec instance to the ´dev-conatiner-server.ts´ which forwads all communcation to stdio.

Support for having multiple devcontainer files is also added

I would love some feedback on this, especially what missing features would be most important to make it most useful.

#### How to test

You need docker installed and started on your machine

1. open a folder in theia
2. create a new `.devcontainer` folder at the root
3. add a `devcontainer.json` and neccessary resources for it
here is an example file using a image wich is automaticly pulled by docker
```
{
    "name": "test_container",
    "image": "nikolaik/python-nodejs:python3.11-nodejs20-slim",
    "forwardPorts": [4000, "3000:3003"],
    "mounts": [{"source": "some/path", "target": "/workspaces", "type": "bind" }]

}

```
or here is a dockerfile with its respective devcontainer.json file for testing dockerfile support:
[devcontainer.json](https://github.com/eclipse-theia/theia/files/14236525/devcontainer.json)
[Dockerfile.txt](https://github.com/eclipse-theia/theia/files/14236526/Dockerfile.txt)

4. execute `Dev Containers: Reopen in Container` 
5. your workspace should then open inside the container

#### Follow-ups

Things not implemented yet which vscodes supports:
- Port forwarding at runtime (this feature is in general still missing in the remote feature) (its mostly done though, there will be a PR as soon as this one is merged)
- attaching to running containers
- Docker Compose support
- variables inside `dockerconfig.json`
- Showing the docker ouput, especially for when images are build through dockerfiles or pulled
- lots of smaller configuration options

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/dev-container/src/electron-node/remote-container-connection-provider.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        const deferred = new Deferred<RemoteExecResult>();
+        try {
+            // TODO add windows container support
+            const execution = await this.container.exec({ Cmd: ['sh', '-c', `${cmd} ${args?.join(' ') ?? ''}`], AttachStdout: true, AttachStderr: true });
+            let stdoutBuffer = '';
+            let stderrBuffer = '';
+            const stream = await execution?.start({});
+            stream.on('close', () => {
+                if (deferred.state === 'unresolved') {
+                    deferred.resolve({ stdout: stdoutBuffer.toString(), stderr: stderrBuffer.toString() });
```
Comment (msujew) [inline packages/dev-container/src/electron-node/remote-container-connection-provider.ts:217]: Minor:
```suggestion
                    deferred.resolve({ stdout: stdoutBuffer, stderr: stderrBuffer });
```

File: packages/core/src/browser/window/window-service.ts [hunk 2]
```diff
@@ -35,7 +35,7 @@ export interface WindowService {
      * Opens a new default window.
      * - In electron and in the browser it will open the default window without a pre-defined content.
      */
-    openNewDefaultWindow(params?: WindowSearchParams): void;
+    openNewDefaultWindow(params?: { search?: WindowSearchParams, hash?: string }): void;
```
Comment (msujew) [inline packages/core/src/browser/window/window-service.ts:38]: Suggestion: We should have a proper type for this. Something like:

```ts
export interface WindowOptions {
  search?: WindowSearchParams
  hash?: string
}
```

```suggestion
    openNewDefaultWindow(options?: { search?: WindowSearchParams, hash?: string }): void;
```

File: packages/core/src/electron-browser/window/electron-window-service.ts [hunk 3]
```diff
@@ -22,6 +22,10 @@ import { ElectronWindowPreferences } from './electron-window-preferences';
 import { ConnectionCloseService } from '../../common/messaging/connection-management';
 import { FrontendIdProvider } from '../../browser/messaging/frontend-id-provider';
 
+export interface WindowReloadOptions {
+    search?: WindowSearchParams,
+    hash?: string
+}
```
Comment (msujew) [inline packages/core/src/electron-browser/window/electron-window-service.ts:28]: Suggestion: Put those to the interface definition.

File: packages/dev-container/src/electron-node/remote-container-connection-provider.ts [hunk 4]
```diff
@@ -159,13 +163,20 @@ export class RemoteDockerContainerConnection implements RemoteConnection {
     }
 
     async forwardOut(socket: Socket): Promise<void> {
-        if (!this.containerInfo) {
-            this.containerInfo = await this.container.inspect();
+        const node = `${this.remoteSetupResult.nodeDirectory}/bin/node`;
```
[the remark this answers is not in the document]
Comment (jonah-iden) [reply packages/dev-container/src/electron-node/remote-container-connection-provider.ts:166]: nope, according to the vscode docs only linux containers are currently supported

File: packages/dev-container/src/dev-container-server/dev-container-server.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+let backendPort: number | undefined = undefined;
+argv.slice(2).forEach(arg => {
+    if (arg.startsWith('-target-port')) {
+        backendPort = parseInt(arg.split('=')[1]);
+    }
+});
+
+if (!backendPort) {
+    throw new Error('please start with -target-port={port number}');
+}
```
[the remark this answers is not in the document]
Comment (jonah-iden) [reply packages/dev-container/src/dev-container-server/dev-container-server.ts:34]: oh thats the port the communiciation is forwarded to. So currently always the port the theia backend is running on.
This dev-container-server is not really listening on any port. only on stdio

**Comment by JonasHelming** [comment]
@msujew : Once you give your approval, we will do another test, does this work for you?

**Comment by jonah-iden** [comment]
@JonasHelming Since Mark is on vacation for the rest of the Month, feel free to test it and give feedback if you find the time

File: packages/dev-container/src/electron-node/dev-container-file-service.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                fs.readdirSync(devcontainerPath.join(filename).fsPath()).map(inner => devcontainerPath.join(filename).join(inner).fsPath()) :
+                [devcontainerPath.join(filename).fsPath()];
+        })
+            .filter(file => file.endsWith('devcontainer.json'))
+            .map(file => ({
+                name: parse(fs.readFileSync(file, 'utf-8')).name ?? 'devcontainer',
+                path: file
+            }));
+
+    }
```
Comment (planger) [inline packages/dev-container/src/electron-node/dev-container-file-service.ts:62]: I find the code for identifying the devcontainer file a bit hard to follow. It'd probably help to assign intermediate results into named variables before chaining them further, and extract certain more complex functions out in separate methods/functions.

File: packages/dev-container/src/electron-node/devcontainer-file.ts [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+    workspaceFolder?: string
+    /**
+     * The --mount parameter for docker run. The default is to mount the project folder at /workspaces/$project.
+     */
+    workspaceMount?: string
+    [k: string]: unknown
+}
+
+// NOT SUPPORTED YET
+// export interface ComposeContainer {
```
Comment (planger) [inline packages/dev-container/src/electron-node/devcontainer-file.ts:140]: I'm inclined to suggest we make the interface definition already complete and only add a comment or so what's supported or not. This makes it easier to complete it step by step, also there are other properties already defined in other interfaces that aren't supported too.

File: packages/dev-container/src/electron-node/remote-container-connection-provider.ts [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+            socket.pipe(stream);
+            ttySession.modem.demuxStream(stream, socket, socket);
+        } catch (e) {
+            console.error(e);
+        }
+    }
+
+    async exec(cmd: string, args?: string[], options?: RemoteExecOptions): Promise<RemoteExecResult> {
+        // return (await this.getOrCreateTerminalSession()).executeCommand(cmd, args);
```
Comment (planger) [inline packages/dev-container/src/electron-node/remote-container-connection-provider.ts:189]:
```suggestion
```

**Review by planger — COMMENTED** [review]
Thank you very much @jonah-iden! This is great work and already works pretty nicely.

My biggest wish would be to show the output of the docker container, to help diagnosing issues. Currently it is very hard to debug situations where the container doesn't start for some reasons, which may happen quite easily since we don't support the full devcontainer format yet.

Aside from that I have a security related question: Afais we don't protect the exposed port of the Theia backend running in the container from the outside world, or did I miss that?
If we don't, I suggest to bind at least the Theia backend port only to the local IP address by default to avoid having it exposed to the entire network.

Other than that I only have a few minor nitpicks inline.

Thanks again for your great work on this feature! This is truly a game-changer, and I'm eager to see it evolved incrementally to support more and more properties of the devcontainer format.

**Comment by jonah-iden** [comment]
Thanks for the review! I'll try to add showing the output today

**Comment by jonah-iden** [comment]
@planger added a for now very basic output display. There is some messages not looking super nice but that can be improved later on

**Comment by JonasHelming** [comment]
@jonah-iden When this is ready for re-reviewe, please just trigger Philips review again

**Review by planger — APPROVED** [review]
Thank you for the additions @jonah-iden! This looks great and I think it should be good to go as a first version of this excellent feature.

**Comment by hsnoil** [comment]
@jonah-iden for this patch, are ENV variables like ${localWorkspaceFolder} and ${localEnv:HOME} suppose to work for mounting? Because it keeps saying add absolute path

https://code.visualstudio.com/remote/advancedcontainers/add-local-file-mount
